File: packages/core/src/snapshot.ts
import type { UserContext, ComputedIslamicDay } from './types.js';
import { getIslamicDayHijriDate } from './calendar.js';
import { getPrayerTimesForDate, addDays } from './prayer-times.js';
import { buildTimeline } from './day-bounds.js';
import { getCurrentPhase, getNextTransition } from './phases.js';
import { getCountdown, getCountdownTarget } from './countdown.js';
import { getIslamicDayProgress, getMarkers, getRingSegments } from './ring.js';

/**
 * Main orchestrator: computes the full Islamic day snapshot
 * from the user's current time, location, and timezone.
 */
export function computeIslamicDaySnapshot(input: UserContext): ComputedIslamicDay {
  const { now, location } = input;

  const todayPT = getPrayerTimesForDate(now, location);
  const yesterdayPT = getPrayerTimesForDate(addDays(now, -1), location);
  const tomorrowPT = getPrayerTimesForDate(addDays(now, 1), location);

  const timeline = buildTimeline(now, todayPT, yesterdayPT, tomorrowPT);

  const hijriDate = getIslamicDayHijriDate(now, todayPT.maghrib);

  const currentPhase = getCurrentPhase(now, timeline);
  const nextTransition = getNextTransition(now, timeline);
  const countdownTarget = getCountdownTarget(now, timeline);
  const countdownMs = getCountdown(now, countdownTarget);

  const progress = getIslamicDayProgress(now, timeline.lastMaghrib, timeline.nextMaghrib);
  const markers = getMarkers(timeline);
  const segments = getRingSegments(timeline);

  return {
    hijriDate,
    prayerTimes: {
      fajr: timeline.fajr,
      sunrise: timeline.sunrise,
      dhuhr: timeline.dhuhr,
      asr: timeline.asr,
      maghrib: timeline.lastMaghrib,
      isha: timeline.isha,
    },
    derivedMarkers: {
      islamicMidnight: timeline.islamicMidnight,
      lastThirdStart: timeline.lastThirdStart,
    },
    timeline,
    currentPhase,
    nextTransition,
    countdownMs,
    ring: {
      progress,
      markers,
      segments,
    },
  };
}

File: apps/apple-watch/IslamicDayShared/IslamicDayCore/Snapshot.swift
import Foundation

/// Main orchestrator: computes full Islamic day snapshot.
/// Mirrors packages/core/src/snapshot.ts

public func computeIslamicDaySnapshot(
    now: Date = Date(),
    location: Location = .mecca,
    timezone: TimeZone = .current
) -> ComputedIslamicDay? {
    guard let todayPT = getPrayerTimesForDate(date: now, location: location),
          let yesterdayPT = getPrayerTimesForDate(date: addDays(date: now, days: -1), location: location),
          let tomorrowPT = getPrayerTimesForDate(date: addDays(date: now, days: 1), location: location) else {
        return nil
    }
    
    let timeline = buildTimeline(
        now: now,
        todayPT: todayPT,
        yesterdayPT: yesterdayPT,
        tomorrowPT: tomorrowPT,
        location: location
    )
    
    let hijriDate = getIslamicDayHijriDate(now: now, todayMaghrib: todayPT.maghrib)
    let currentPhase = getCurrentPhase(now: now, timeline: timeline)
    let (nextId, nextAt) = getNextTransition(now: now, timeline: timeline)
    let countdownTarget = getCountdownTarget(now: now, timeline: timeline)
    let countdownMs = Int64(max(0, countdownTarget.timeIntervalSince(now) * 1000))
    let progress = getIslamicDayProgress(now: now, lastMaghrib: timeline.lastMaghrib, nextMaghrib: timeline.nextMaghrib)
    let markers = getMarkers(timeline: timeline)
    let segments = getRingSegments(timeline: timeline)
    
    return ComputedIslamicDay(
        hijriDate: hijriDate,
        prayerTimes: todayPT,
        timeline: timeline,
        currentPhase: currentPhase,
        nextTransitionId: nextId,
        nextTransitionAt: nextAt,
        countdownMs: countdownMs,
        ringProgress: progress,
        ringMarkers: markers,
        ringSegments: segments
    )
}
